### ISSUE TITLE
[FA3] `nan` gradients in the BW pass

### ISSUE THREAD

**Issue by danthe3rd** [Dao-AILab/flash-attention#1052]
Hi,
I'm getting some nans in the BW pass, here is a sample repro:
(NOTE: The `* 3` scaling factor is necessary to get the nans)
```python
import torch
import flashattn_hopper_cuda
from flash_attn_interface import flash_attn_func
 
M = 1024
 
torch.set_default_device("cuda")
torch.set_default_dtype(torch.float16)
 
# direct calls to API
query = torch.randn([1, M, 1, 128]) * 3
key = torch.randn([1, M, 1, 128]) * 3
value = torch.randn([1, M, 1, 128]) * 3
dout = torch.randn([1, M, 1, 128])
out = torch.zeros([1, M, 1, 128])
softmax_scale = 128 ** -0.5
is_causal = True
 
out, q_padded, k_padded, v_padded, out_padded, softmax_lse, p = flashattn_hopper_cuda.fwd(query, key, value, out, softmax_scale, is_causal)
assert not out.isnan().any().item()
dq = torch.zeros_like(query)
dk = torch.zeros_like(key)
dv = torch.zeros_like(value)
dq, dk, dv, softmax_d = flashattn_hopper_cuda.bwd(
    dout, query, key, value, out, softmax_lse, dq, dk, dv, softmax_scale, is_causal
)
print(dq)
 
 
# with interface
query.requires_grad_(True)
key.requires_grad_(True)
value.requires_grad_(True)
out = flash_attn_func(query, key, value, softmax_scale=softmax_scale, causal=is_causal)[0]
out.backward(torch.randn_like(out))
print(query.grad)
```

**Comment by ipiszy** [comment]
Hi @danthe3rd , this problem should have been solved with the new bwd kernel commit. I tried your script with `assert not dq.isnan().any().item()` and didn't observe errors. Could you check the latest commit?

**Comment by varunneal** [comment]
Hello @ipiszy do you know if this issue is resolved for `flash_attn_varlen_func`? I am getting severe `nan` gradients using this function.
